src/ServiceControl.Transports.SQLServer/QueueLengthProvider.cs

```diff
@@ -0,0 +1,193 @@
+	using Monitoring;
+	using Monitoring.Infrastructure;
+	using Monitoring.Messaging;
+	using Monitoring.QueueLength;
+	using NServiceBus.Logging;
+	using NServiceBus.Metrics;
+
+	public class QueueLengthProvider : IProvideQueueLength
+	{
+		ConcurrentDictionary<EndpointInputQueue, SqlTable> endpointInputQueueToTableName = new ConcurrentDictionary<EndpointInputQueue, SqlTable>();
+		ConcurrentDictionary<SqlTable, int> tableSizes = new ConcurrentDictionary<SqlTable, int>();
+
+		string connectionString;
+		QueueLengthStore store;
+
+		CancellationTokenSource stop = new CancellationTokenSource();
+		Task pooler;
+
+		public void Initialize(string connectionString, QueueLengthStore store)
+		{
+			this.connectionString = connectionString;
+			this.store = store;
+		}
+
+		public void Process(EndpointInstanceId endpointInstanceId, EndpointMetadataReport metadataReport)
+		{
+			var endpointInputQueue = new EndpointInputQueue(endpointInstanceId.EndpointName, metadataReport.LocalAddress);
+			var localAddress = metadataReport.LocalAddress;
+			var pluginVersion = metadataReport.PluginVersion;
+
+			if (SqlTable.TryParse(localAddress, pluginVersion, out var sqlTable))
+			{
+			    SqlTable previousSqlTable = null;
+
+				endpointInputQueueToTableName.AddOrUpdate(endpointInputQueue, _ => sqlTable, (_, cv) =>
+				{
+					previousSqlTable = cv;
+					return sqlTable;
+				});
+
+			    if (previousSqlTable == null)
+			    {
+			        tableSizes.TryAdd(sqlTable, 0);
+                }
+				else if (previousSqlTable.Equals(sqlTable) == false)
+				{
+					tableSizes.TryRemove(previousSqlTable, out var _);
+					tableSizes.TryAdd(sqlTable, 0);
+				}
+			}
+			else
+			{
+				Logger.Warn($"Could not localAddress {localAddress}. Sent by {endpointInstanceId.EndpointName} id {endpointInstanceId.InstanceId}. Plugin version {pluginVersion}");
+			}
+		}
+
+		public void Process(EndpointInstanceId endpointInstanceId, TaggedLongValueOccurrence metricsReport)
+		{
+			//HINT: Sql server endpoints do not support endpoint level queue length monitoring
+		}
+	   
+		public Task Start()
+		{
+			stop = new CancellationTokenSource();
+
+			pooler = Task.Run(async () =>
+			{
+				while (!stop.Token.IsCancellationRequested)
+				{
+					try
+					{
+						await QueryTableSizes();
+
+						UpdateQueueLengthStore();
+
+						await Task.Delay(QueryDelayInterval);
+					}
+					catch (Exception e)
+					{
+						Logger.Error("Error querying sql queue sizes.", e);
+					}
+				}
+			});
+
+			return TaskEx.Completed;
+		}
+
+	    public Task Stop()
+	    {
+	        stop.Cancel();
+
+	        return pooler;
+	    }
+
+        void UpdateQueueLengthStore()
+		{
+			var nowTicks = DateTime.UtcNow.Ticks;
+
+			foreach (var endpointToTableNamePair in endpointInputQueueToTableName)
+			{
+				store.Store(
+					new []{ new RawMessage.Entry
+					{
+						DateTicks = nowTicks,
+						Value = tableSizes.TryGetValue(endpointToTableNamePair.Value, out var size) ? size : 0
```

Comment:
Good to see `TryGetValue` in here. We've got two concurrent dictionaries accessed concurrently.